optimizely/decision_service.py

```diff
@@ -323,128 +329,216 @@ def get_variation(
       array of log messages representing decision making.
     """
         decide_reasons = []
-        # Go through each experiment in order and try to get the variation for the user
-        if rollout and len(rollout.experiments) > 0:
-            for idx in range(len(rollout.experiments) - 1):
-                logging_key = str(idx + 1)
-                rollout_rule = project_config.get_experiment_from_id(rollout.experiments[idx].get('id'))
-
-                # Check if user meets audience conditions for targeting rule
-                audience_conditions = rollout_rule.get_audience_conditions_or_ids()
-                user_meets_audience_conditions, reasons_received = audience_helper.does_user_meet_audience_conditions(
-                    project_config,
-                    audience_conditions,
-                    enums.RolloutRuleAudienceEvaluationLogs,
-                    logging_key,
-                    attributes,
-                    self.logger)
+
+        if not feature:
+            return Decision(None, None, enums.DecisionSources.ROLLOUT), decide_reasons
+
+        rollout = project_config.get_rollout_from_id(feature.rolloutId)
+
+        if not rollout:
+            message = 'There is no rollout of feature {}.'.format(feature.key)
+            self.logger.debug(message)
+            decide_reasons.append(message)
+            return Decision(None, None, enums.DecisionSources.ROLLOUT), decide_reasons
+
+        rollout_rules = project_config.get_rollout_experiments(rollout)
+
+        if not rollout_rules:
+            message = 'Rollout {} has no experiments.'.format(rollout.id)
+            self.logger.debug(message)
+            decide_reasons.append(message)
+            return Decision(None, None, enums.DecisionSources.ROLLOUT), decide_reasons
+
+        if rollout and len(rollout_rules) > 0:
+            index = 0
+            while index < len(rollout_rules):
+                decision_response, reasons_received = self.get_variation_from_delivery_rule(project_config,
+                                                                                            feature,
+                                                                                            rollout_rules, index, user,
+                                                                                            options)
+
                 decide_reasons += reasons_received
-                if not user_meets_audience_conditions:
-                    message = 'User "{}" does not meet conditions for targeting rule {}.'.format(user_id, logging_key)
-                    self.logger.debug(
-                        message
-                    )
-                    decide_reasons.append(message)
-                    continue
-                message = 'User "{}" meets audience conditions for targeting rule {}.'.format(user_id, idx + 1)
+
+                if decision_response:
+                    variation, skip_to_everyone_else = decision_response
```

Comment:
Do we have a case for None decision_response? Remove if or add else.

Reply:
Done